internal/plugin/connectors/tcp/mssql/connector_test.go

```diff
@@ -1,59 +1,372 @@
 package mssql
 
 import (
-	"context"
+	"errors"
+	"io"
 	"net"
 	"testing"
 
+	errorspkg "github.com/pkg/errors"
 	"github.com/stretchr/testify/assert"
 
 	"github.com/cyberark/secretless-broker/internal/plugin/connectors/tcp/mssql/mock"
-	logmock "github.com/cyberark/secretless-broker/pkg/secretless/log/mock"
+	"github.com/cyberark/secretless-broker/internal/plugin/connectors/tcp/mssql/types"
 	pluginconnector "github.com/cyberark/secretless-broker/pkg/secretless/plugin/connector"
 	"github.com/denisenkom/go-mssqldb"
 )
 
-func TestHappyPath(t *testing.T) {
-	logger := logmock.NewLogger()
-	clientConn := mock.NewNetConn(nil)
+func TestConnectSuccess(t *testing.T) {
 	expectedBackendConn := mock.NewNetConn(nil)
-	creds := pluginconnector.CredentialValuesByID{
-		"credName": []byte("secret"),
+
+	connector := NewSingleUseConnectorWithOptions(
+		mock.DefaultConnectorOptions(),
+		mock.MSSQLConnectorCtor(
+			func(options *mock.MSSQLConnectorCtorOptions) {
+				options.BackendConn = expectedBackendConn
+			},
+		),
+	)
+	actualBackendConn, err := runDefaultTestConnect(connector)
+
+	assert.Nil(t, err)
+	if err != nil {
+		return
+	}
+
+	assert.Equal(t, expectedBackendConn, actualBackendConn)
+}
+
+func TestProductionReadPreLoginRequest(t *testing.T) {
+	// production version of ReadPreLoginRequest
+	var readPreLoginRequest types.ReadPreloginRequestFunc = mssql.ReadPreloginRequest
+
+	// expected prelogin request available from net.Conn passed to ReadLogin
+	expectedPreLoginRequest := map[uint8][]byte{
+		1: {2,3,4},
+	}
+
+	r, w := net.Pipe()
+	go func() {
+		_ = mssql.WritePreloginRequest(w, expectedPreLoginRequest)
+	}()
+
+	// prelogin request returned from ReadPreLoginRequest
+	actualLoginRequest, _ := readPreLoginRequest(r)
+
+	assert.Equal(t, actualLoginRequest, expectedPreLoginRequest)
+}
+
+func TestReadPreLoginRequestFails(t *testing.T) {
+	methodFails(t, func(connectorOptions *types.ConnectorOptions) {
+		connectorOptions.ReadPreloginRequest = func(
+			io.ReadWriteCloser,
+		) (map[uint8][]byte, error) {
+			return nil, methodFailsExpectedErr
```

Comment:
The methodFails method depends upon (or assumes that) the test that calls provides function options that must return 'methodFailsExpectedErr'. This does minimize the code, but do we want to keep this implicit dependency for methodFails, or alternatively pass the expected error to methodFails?